sphinxcontrib/mat_tree_sitter_parser.py

```diff
@@ -0,0 +1,940 @@
+    else:
+        return point.row
+
+
+def process_text_into_docstring(text, encoding):
+    """Take a text bytestring and decode it into a docstring."""
+    docstring = text.decode(encoding, errors="backslashreplace")
+    return re.sub(re_percent_remove, "", docstring)
+
+
+def process_default(node, encoding):
+    """Take the node defining a default and remove any line continuations before generating the default."""
+    text = node.text
+    to_keep = set(range(node.end_byte - node.start_byte))
+    lc_matches = q_line_continuation.matches(node)
+    for _, match in lc_matches:
+        # TODO this copies a lot perhaps there is a better option.
+        lc = match["lc"]
+        cut_start = lc.start_byte - node.start_byte
+        cut_end = lc.end_byte - node.start_byte
+        to_keep -= set(range(cut_start, cut_end))
+    # NOTE: hardcoded endianess is fine because for one byte this does not matter.
+    #       See python bikeshed on possible defaults for this here:
+    #       https://discuss.python.org/t/what-should-be-the-default-value-for-int-to-bytes-byteorder/10616
+    new_text = b"".join(
+        [byte.to_bytes(1, "big") for idx, byte in enumerate(text) if idx in to_keep]
+    )
+    # TODO We may want to do an in-order traversal of the parse here to generate a "nice" reformatted single line
+    #      however doing so sufficiently generically is likely a major undertaking.
+    default = new_text.decode(encoding, errors="backslashreplace")
+    default = re.sub(re_assign_remove, "", default)
+    return re.sub(re_trim_line, "", default)
+
+
+class MatScriptParser:
+    def __init__(self, root_node, encoding):
+        """Parse m script"""
+        self.encoding = encoding
+        script_matches = q_script.matches(root_node)
+        if script_matches:
+            _, script_match = q_script.matches(root_node)[0]
+            docstring_node = script_match.get("docstring")
+            if docstring_node is not None:
+                self.docstring = process_text_into_docstring(
+                    docstring_node.text, self.encoding
+                )
+            else:
+                self.docstring = None
+        else:
+            self.docstring = None
+
+
+class MatFunctionParser:
+    def __init__(self, root_node, encoding):
+        """Parse Function definition"""
+        self.encoding = encoding
+        _, fun_match = q_fun.matches(root_node)[0]
+        self.name = fun_match.get("name").text.decode(
+            self.encoding, errors="backslashreplace"
+        )
+
+        # Get outputs (possibly more than one)
+        self.retv = {}
+        output_nodes = fun_match.get("outputs")
+        if output_nodes is not None:
+            retv = [
+                output.text.decode(self.encoding, errors="backslashreplace")
+                for output in output_nodes
+            ]
+            for output in retv:
+                self.retv[output] = {}
+
+        # Get parameters
+        self.args = {}
```

Comment:
Why is this a dictionary, if the value is just an empty dictionary. Do you plan on storing more information in there.
The reason I'm asking, is that it makes the downstream usage more complicated.

Reply:
Similarly to retv this is a dictionary as it allows for the storage of property validators in an `arguments` block!